awx/main/scheduler/task_manager.py

```diff
@@ -249,20 +272,47 @@ def timeout_approval_node(self):
     def get_tasks(self):
         workflow_jobs_running = [wf for wf in WorkflowJob.objects.filter(status='running')]
         workflow_jobs_pending = [wf for wf in WorkflowJob.objects.filter(status='pending')]
+        return workflow_jobs_running, workflow_jobs_pending
+
+    def start_pending_workflows(self, workflow_jobs_running, workflow_jobs_pending):
+        """Take in list of currently pending and running workflows.
+
+        Start the workflows that we can and send any notifications or websockets needed.
+        Return list of the now running workflows.
+        """
         workflow_to_start = []
-        running_workflow_pk = {wf.pk for wf in workflow_jobs_running}
+        running_wfjt_ids = {wf.unified_job_template_id for wf in workflow_jobs_running}
         for wf in workflow_jobs_pending:
-            if wf.allow_simultaneous or wf.pk not in running_workflow_pk:
+            if wf.allow_simultaneous or wf.unified_job_template_id not in running_wfjt_ids:
                 wf.status = 'running'
                 workflow_to_start.append(wf)
+                running_wfjt_ids.add(wf.unified_job_template_id)
+                logger.debug('Transitioning %s to running status.', wf.log_format)
+                self.start_task_limit -= 1
+                if self.start_task_limit == 0:
+                    break
+                if self.timed_out():
+                    logger.warning(f"Workflow manager has reached time out processing pending workflows, exiting loop early")
+                    break
+            else:
+                logger.debug('Workflow %s staying in pending, blocked by another running workflow from the same workflow job template', wf.log_format)
 
         WorkflowJob.objects.bulk_update(workflow_to_start, ['status'])
+
+        # We could do the following in the previous loop
+        # Doing it in a loop after the bulk_update preserves previous behavior
+        # of not sending the notifications until after the status of running has been saved.
+        for wf in workflow_to_start:
+            wf.send_notification_templates('running')
+            wf.websocket_emit_status(wf.status)
```

Comment:
This is done in a transaction, and you can have race conditions if these actions aren't delayed as `on_commit` methods. But again, this is starting to feel like re-inventing the wheel when the task manager already does this stuff more correctly.